### ISSUE TITLE
Enrich wake's Error type

### ISSUE THREAD

**Issue by JakeSiFive** [sifive/wake#1385]
Right now wake has the following functions for working with errors

- failWithError
- makeError
- addErrorContext

The error type itself looks like this

```
export tuple Error =
    export Cause: String
    export Stack: List String
```

I'd like to propose the following instead

```
export data ErrorTree =
    ErrorList (List Error) # Allows nesting of sub errors
    BaseError (stack: List String) # This acts like the old Error type
    JobError Job # This is a new error type for rich job failures
 
export data Error =
    export Cause: String
    export Printer: (cause: String) => ErrorTree => (nestingLevel: Integer) => Unit
    export Tree: ErrorTree
```

You can then view `Error` as a tree of errors, additionally each part of the tree can render itself as it would like (but is given the nestingLevel that should be respected so some ad hoc standard should be devoted to that).

The follow up to this proposal is that Wake runs the Printer code if a `Result a Error` is returned instead of printing out `Fail {format error}`

**Comment by JakeSiFive** [comment]
@bmitc @V-FEXrt @mmjconolly @ag-eitilt @richardxia @jackkoenig @mwachs5 

Pinging you all as no one really has anytime to implement this right now as far as I'm aware but y'all all might catch a wild hair and want to make this sketch of error handling a reality.

**Comment by richardxia** [comment]
I'm afraid your proposed data structures are pretty confusing to me. It's a little difficult for me to express why that is the case, but I think it's related to your specific choices of which piece of data go into which of the two mutually recursive types.

I feel that a more intuitive set of data structures to me would either 1) collapse everything into a single data type or it would 2) fully separate the recursive structure from the domain-specific data. Here are some concrete examples of what I'm thinking:

```wake
# Single ADT

export tuple Error =
  export Cause: String
  export Printer: # Something goes here
  export Stack: List String
  export Job: Option Job
  export Children: Error


# Fully separated recursive structure

export data Tree a =
  Tree a (List Tree)
  Leaf a

export tuple ErrorData =
  export Cause: String
  export Printer: # Something goes here
  export Stack: List String
  export Job: Option Job

# Since we can't write type Error = Tree ErrorData, wrap Tree ErrorData with a tuple
tuple Error:
  Tree: Tree ErrorData
```

After writing the above out, I think maybe the biggest issue I have with your proposed structure is that it splits the stack trace from the `Error` tuple, which means that an Error that wraps other Errors cannot simultaneously have a stack trace, but I'm pretty sure it'll be useful to understand exactly when the wrapper Error was created.

**Comment by richardxia** [comment]
> Your single ADT option looks like it solves that to me but I'd rather children be a list (which you might have already intended given the name you used)

Yeah, sorry, typo on my part. But yeah, I like your revised data structure.
